pallets/asset/src/checkpoint.rs

```diff
@@ -0,0 +1,551 @@
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+//! # Checkpoint Module
+//!
+//! The Checkpoint module provides extrinsics and storage to take snapshots,
+//! henceforth called *checkpoints*, of the supply of assets,
+//! and how they were distributed at the time of checkpoint.
+//!
+//! Using the module, users can also schedule checkpointing in the future,
+//! either at fixed points in time (e.g., "next friday"),
+//! or at regular intervals (e.g., "every month").
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Functions
+//!
+//! - `create_checkpoint` creates a checkpoint.
+//! - `create_schedule` creates a checkpoint schedule.
+//! - `remove_schedule` removes a checkpoint schedule.
+//!
+//! ### Public Functions
+//!
+//! - `balance_at(ticker, did, cp)` returns the balance of `did` for `ticker` at least `>= cp`.
+//! - `advance_update_balances(ticker, updates)` advances schedules for `ticker`
+//!    and applies new balances in `updates` for the last checkpoint.
+//! - Other misc storage items as defined in `decl_storage!`.
+
+use codec::{Decode, Encode};
+use core::mem;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage,
+    dispatch::{DispatchError, DispatchResult},
+    ensure,
+    traits::{Get, UnixTime},
+};
+use frame_system::ensure_root;
+use polymesh_common_utilities::{
+    protocol_fee::{ChargeProtocolFee, ProtocolOp},
+    CommonTrait, GC_DID,
+};
+use polymesh_primitives::{
+    calendar::{CheckpointId, CheckpointSchedule},
+    IdentityId, Moment, Ticker,
+};
+use sp_runtime::traits::SaturatedConversion;
+#[cfg(feature = "std")]
+use sp_runtime::{Deserialize, Serialize};
+use sp_std::prelude::*;
+
+use crate as pallet_asset;
+use pallet_asset::Trait;
+
+type Asset<T> = pallet_asset::Module<T>;
+
+/// ID of a `StoredSchedule`.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Copy, Clone, Debug, PartialEq, Eq, Default)]
+pub struct ScheduleId(pub u64);
+
+/// One or more scheduled checkpoints in the future.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Copy, Clone, Debug, PartialEq, Eq)]
+pub struct StoredSchedule {
+    /// A series of checkpoints in the future defined by the schedule.
+    pub schedule: CheckpointSchedule,
+    /// The ID of the schedule itself.
+    /// Not to be confused for a checkpoint's ID.
+    pub id: ScheduleId,
+    /// When the checkpoint is next due to be created.
+    /// Used as a cache for more efficient sorting.
+    pub at: Moment,
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Checkpoint {
+        // --------------------- Supply / Balance storage ----------------------
+
+        /// Total supply of the token at the checkpoint.
+        ///
+        /// (ticker, checkpointId) -> total supply at given checkpoint
+        pub TotalSupply get(fn total_supply_at):
+            map hasher(blake2_128_concat) (Ticker, CheckpointId) => T::Balance;
```

Comment:
Could this be `double_map`?

Reply:
In the standard access pattern for checkpoints you already know the ID, and you can also get the total number of IDs as well as which checkpoints balances for a user were changed in. Switching to a double map would slow things down considerably, and I'm not sure there are uses-cases that need to be satisfied to justify this change?

Reply:
I agree that in general a double map is slower than a single map with keys being pairs. To be fair, if one of the keys is hashed with a fast hash or even `identity` then the slowdown is negligible.